File: boto/leo.py
#!/usr/bin/python
# coding: utf8

from boto.s3.connection import S3Connection, OrdinaryCallingFormat
from boto.s3.bucket import Bucket
from boto.s3.key import Key
from filechunkio import FileChunkIO
import boto
import hashlib
import os
import mimetypes
import random
import traceback
import magic
import math

AWS_ACCESS_KEY = "05236"
AWS_SECRET_ACCESS_KEY = "802562235"
BUCKET_NAME = "test" + str(random.randint(1,99999))  ## Dynamic BucketName
FILE_NAME = "testFile"
CHUNK_SIZE = 5 * 1024 * 1024
LARGE_FILE_PATH = "../temp_data/testFile.large"
LARGE_OBJ_SIZE = 52428800

conn = S3Connection(AWS_ACCESS_KEY,
    AWS_SECRET_ACCESS_KEY,
    host = "localhost",
    port = 8080,
    calling_format = OrdinaryCallingFormat(),
    is_secure = False
)

try:
    print BUCKET_NAME
    # Create bucket
    print "Bucket Creation Test [Start]\n"
    buckets = conn.create_bucket(BUCKET_NAME)
    print "Bucket Created Successfully"

    # Show buckets
    print "--------Bucket List------"
    for bucket in conn.get_all_buckets():
        print bucket
    print "Bucket Creation Test [End]\n"

    # Get Bucket
    bucket = conn.get_bucket(BUCKET_NAME,validate=False)
    print "Get Bucket Successfully\n"

    # Put Object
    print "File Upload Test [Start]\n"
    file_path = "../temp_data/" + FILE_NAME
    file_object = open(file_path, "r")
    file_digest = hashlib.md5(file_object.read()).hexdigest()
    file_size = os.path.getsize(file_path)
    file_type = magic.from_file(file_path, mime=True)

    # Files in Amazon S3 are called "objects" and are stored in buckets. A specific object is 
    # referred to by its key (i.e., name) and holds data. Here, we create a new object with 
    # the key name, HEAD request is Metadata of that object. e.g. Size, etag, Content_type etc.
    # For more information http://boto.readthedocs.org/en/latest/s3_tut.html#storing-data

    # PUT Object using single-part method
    print "File is being upload:"
    bucket.new_key(FILE_NAME).set_contents_from_filename(file_path)

    if not os.path.exists(LARGE_FILE_PATH):
        with open(LARGE_FILE_PATH, "wb") as f:
            f.write(os.urandom(LARGE_OBJ_SIZE))
    # Put Single-Part Large Object
    bucket.new_key(FILE_NAME + ".large.one").set_contents_from_filename(LARGE_FILE_PATH)

#    # Put Multi-Part Large Object
#    multi_part = bucket.initiate_multipart_upload(FILE_NAME + ".large.part")
#    chunk_count = int(math.ceil(LARGE_OBJ_SIZE / float(CHUNK_SIZE)))
#    for i in range(chunk_count):
#        offset = CHUNK_SIZE * i
#        bytes = min(CHUNK_SIZE, LARGE_OBJ_SIZE - offset)
#        with FileChunkIO(LARGE_FILE_PATH, 'rb', offset=offset,
#                                         bytes=bytes) as fp:
#            multi_part.upload_part_from_file(fp, part_num=i + 1)
#    multi_part.complete_upload()

    # HEAD Object
    obj = bucket.get_key(FILE_NAME)
    if not(obj.exists()):
        raise "Object doesn't exists"
    if not(file_size == obj.size and file_digest == obj.etag[1:-1]):
        raise "File Metadata could not match"
    else:
        print "File MetaData : Content_type:", obj.content_type, "\b, Content_encoding:", obj.content_encoding
        print "\b, etag:", obj.etag, "\b, Size:", obj.size, "\b, Name:", obj.name, "\n"

    # GET object
    if not file_size == obj.size:
        raise "Upload File content is not equal\n"
    if "text/plain" in file_type:
        print "Uploaded object data : \t", obj.read()
    else:
        print "File Content type is :", obj.content_type

    # GET non-existing object
    nonexist_obj = bucket.get_key(FILE_NAME + ".nonexist")
    try:
        print "Non-existing files:", nonexist_obj.read()
        raise "File must NOT be exist\n"
    except Exception, not_found:
        print "Get non-existing object successfully\n"
    print "Get non-existing object Test [End]\n"

    # Range Get object
    print "Range Get Object Test [Start]";
    print "Range Get in Small Object";
    rangeDict = {'Range': 'bytes=1-4'}
    obj_range = bucket.get_key(FILE_NAME)
    resp = obj_range.get_contents_as_string(headers = rangeDict)
    if resp != "his ":
        raise "Range Get Result does NOT match\n"
    else:
        print "Range Get Succeeded\n"

    with open(LARGE_FILE_PATH, "rb") as f:
        f.seek(1048576)
        base_arr = f.read(10485760 - 1048576 + 1)

    print "Range Get in Single Part Large Object"
    rangeDict = {'Range': 'bytes=1048576-10485760'}
    obj_range = bucket.get_key(FILE_NAME + ".large.one")
    resp = obj_range.get_contents_as_string(headers = rangeDict)
    if base_arr != resp:
        raise "Range Get Result does NOT match\n"
    else:
        print "Range Get Succeeded\n"

#    print "Range Get in Multiple Part Large Object"
#    rangeDict = {'Range': 'bytes=1048576-10485760'}
#    obj_range = bucket.get_key(FILE_NAME + ".large.part")
#    resp = obj_range.get_contents_as_string(headers = rangeDict)
#    if base_arr != resp:
#        raise "Range Get Result does NOT match\n"
#    else:
#        print "Range Get Succeeded\n"

    print "Range Get Object Test [End]\n"

    # Show Objects
    print"--------------------------------List Objects-----------------------------------"
    for key in bucket.list():
        print key.name, "\t\t", key.size, "\t\t", key.last_modified
        if not(file_size == key.size):
            if not(LARGE_OBJ_SIZE == key.size):
                raise "Content length is changed for :", key.size
    print "File Uploaded Successfully\n"
    print "File Upload Test [End]\n"

    # File copy
    print "File Copy Test [Start]\n"
    obj = bucket.copy_key(FILE_NAME + ".copy", BUCKET_NAME, FILE_NAME)
    if not(obj.exists()):
       raise "File could not Copy Successfully\n"

    # Show Objects
    print"--------------------------------List Objects-----------------------------------"
    for key in bucket.list():
        print key.name, "\t\t", key.size, "\t\t", key.last_modified
        if not(file_size == key.size):
            if not(LARGE_OBJ_SIZE == key.size):
                raise "Content length is changed for :", key.size
    print "File copied successfully\n"
    print "File Copy Test [End]\n"

    # File Download
    print "File Download Test [Start]\n"
    this_file_path = FILE_NAME + ".copy"
    obj.get_contents_to_filename(this_file_path)
    this_file_object = open(this_file_path, "r")
    this_file_digest = hashlib.md5(this_file_object.read()).hexdigest()
    this_file_size = os.path.getsize(this_file_path)
    if not(this_file_size == obj.size and this_file_digest == file_digest):
        raise "Downloaded File Metadata could not match\n"
    print "File Downloaded Successfully\n"
    print "File Download Test [End]\n"

    # Delete objects one by one and check if exist
    print "File Delete Test [Start]\n"
    print "--------------------------------Delete Objects---------------------------------"
    for key in bucket.list():
        print key.name, "\tDeleted Successfully", key.delete()

    for key in bucket.list():
        if key.exists():
            raise "\tObject is not Deleted Successfully\n"
        try:
            print "Deleted files:", key.read()
        except Exception, not_found:
            print not_found
            continue
        raise "Deleted Failed\n"
    print "\nFile Delete Test [End]\n"

    # Get-Put ACL
    print "Object ACL Test [Start]"
    print "\n#####Default ACL#####"
    acp = bucket.get_acl()
    print acp
    print "Owner ID :" + acp.owner.id
    print "Owner Display name : " + acp.owner.display_name
    permissions = []
    for grant in acp.acl.grants:
        print "Bucket ACL is :", grant.permission,"\nBucket Grantee URI is :", grant.uri
        permissions.append(grant.permission)
    if not all(x in permissions for x in ["FULL_CONTROL"]):
        raise "Permission is Not full_control"
    else:
        print "Bucket ACL permission is 'private'\n"

    print "########:public_read ACL########"
    bucket.set_acl("public-read")
    acp = bucket.get_acl()
    print "Owner ID :", acp.owner.id
    print "Owner Display name :", acp.owner.display_name
    permissions = []
    for grant in acp.acl.grants:
        print "Bucket ACL is :", grant.permission,"\nBucket Grantee URI is :", grant.uri
        permissions.append(grant.permission)
    if not all(x in permissions for x in ["READ","READ_ACP"]):
        raise "Permission is Not public_read"
    else:
        print "Bucket ACL Successfully changed to 'public-read'\n"

    print "#####:public_read_write ACL#####"
    bucket.set_acl("public-read-write")
    acp = bucket.get_acl()
    print "Owner ID :", acp.owner.id
    print "Owner Display name :", acp.owner.display_name
    permissions = []
    for grant in acp.acl.grants:
        print "Bucket ACL is :", grant.permission,"\nBucket Grantee URI is :", grant.uri
        permissions.append(grant.permission)
    if not all(x in permissions for x in ["READ","READ_ACP","WRITE", "WRITE_ACP"]):
        raise "Permission is Not public_read_write"
    else:
        print "Bucket ACL Successfully changed to 'public-read-write'\n"

    print "##########:private ACL##########"
    bucket.set_acl("private")
    acp = bucket.get_acl()
    print "Owner ID :", acp.owner.id
    print "Owner Display name :", acp.owner.display_name
    permissions = []
    for grant in acp.acl.grants:
        print "Bucket ACL is :", grant.permission,"\nBucket Grantee URI is :", grant.uri
        permissions.append(grant.permission)
    if not all(x in permissions for x in ["FULL_CONTROL"]):
        raise "Permission is Not full_control"
    else:
        print "Bucket ACL Successfully changed to 'private'\n"
    print "Bucket ACL Test [End]\n"
except Exception,e :
    print traceback.format_exc()
finally:
    # Delete Bucket
    print "Bucket Delete Test [Start]\n"
    bucket = conn.get_bucket(BUCKET_NAME,validate=False)
    bucket.delete()
    print "Bucket Deleted Successfully\n"
    print "Bucket Delete Test [End]\n"

File: aws-sdk-ruby/leo.rb
## This code supports "aws-sdk v1.9.5"
require "aws-sdk"
require "content_type"

# set your s3 key and variable
Endpoint = "localhost"
Port = 8080
AccessKeyId = "05236"
SecretAccessKey = "802562235"
FileName = "testFile"
ChunkSize = 5 * 1024 * 1024  ## 5 MB chunk size
Bucket = "test" + rand(99999).to_s  ## Dynamic BucketName

class LeoFSHandler < AWS::Core::Http::NetHttpHandler
  def handle(request, response)
    request.port = ::Port
    super
  end
end

SP = AWS::Core::CredentialProviders::StaticProvider.new(
{
  :access_key_id     => AccessKeyId,
  :secret_access_key => SecretAccessKey
})

AWS.config(
  access_key_id: AccessKeyId,
  secret_access_key: SecretAccessKey,
  s3_endpoint: Endpoint,
  http_handler: LeoFSHandler.new,
  credential_provider: SP,
  s3_force_path_style: true,
  use_ssl: false
)

s3 = AWS::S3.new
begin
  # Create bucket
  s3.buckets.create(Bucket)
  puts "Bucket Created Successfully\n"

  # Get bucket
  bucket = s3.buckets[Bucket]
  puts "Get Bucket Successfully\n\n"

  # PUT Object
  file_path = "../temp_data/" + FileName
  fileObject =  open(file_path)

  # PUT object using single-part method
  obj = bucket.objects[FileName + ".single"].write(file: file_path, content_type: fileObject.content_type)

  # PUT object using multi-part method
  puts "File is being upload:\n"
  counter = fileObject.size / ChunkSize
  uploading_object = bucket.objects[File.basename(fileObject.path)]
  uploading_object.multipart_upload(:content_type => fileObject.content_type.to_s) do |upload|
    while !fileObject.eof?
      puts " #{upload.id} \t\t #{counter} "
      counter -= 1
      upload.add_part(fileObject.read ChunkSize) ## 20MB Default size is 5242880 Byte
      p("Aborted") if upload.aborted?
    end
  end
  puts "File Uploaded Successfully\n\n"

  # List objects in the bucket
  puts "----------List Files---------\n"
  bucket.objects.with_prefix("").each do |obj|
    if !fileObject.size.eql? obj.content_length
       raise " Content length is changed for : #{obj.key}"
    end
    puts "#{obj.key} \t #{obj.content_length}"
  end

  # HEAD object
  fileObject.seek(0)
  fileDigest = Digest::MD5.hexdigest(fileObject.read)
  metadata = bucket.objects[FileName + ".single"].head
  if !((fileObject.size.eql? metadata.content_length) && (fileDigest.eql? metadata.etag.gsub('"', ''))) ## for future use  && (fileObject.content_type.eql? metadata.content_type))
    raise "Single Part File Metadata could not match"
  else
    puts "Single Part File MetaData :"
    p metadata
  end
  metadata = bucket.objects[FileName].head
  if !(fileObject.size.eql? metadata.content_length)  ## for future use && (fileObject.content_type.eql? metadata.content_type)
    raise "Multipart File Metadata could not match"
  else
    puts "Multipart Part File MetaData :"
    p metadata
  end

  # GET object(To be handled at the below rescue block)
  if !fileObject.size.eql?  bucket.objects[FileName + ".single"].head.content_length
    raise "\nSignle part Upload File content is not equal\n"
  end
  puts "\nSingle Part Upload object data :\t" + bucket.objects[FileName + ".single"].read
  if !fileObject.size.eql? bucket.objects[FileName].head.content_length
    raise "Multi Part Upload File content is not equal\n"
  end
  if fileObject.content_type.eql? "text/plain"
    puts "Multi Part Upload object data :\t" +  bucket.objects[FileName].read + "\n"
  else
    puts "File Content type is :" + bucket.objects[FileName].content_type + "\n\n"
  end

  # Copy object
  bucket.objects[FileName + ".copy"].copy_from(FileName)
  if !bucket.objects[FileName + ".copy"].exists?
    raise "File could not Copy Successfully\n"
  end
  puts "File copied successfully\n"

  # List objects in the bucket
  puts "----------List Files---------\n"
  bucket.objects.with_prefix("").each do |obj|
    if !fileObject.size.eql? obj.content_length
       raise " Content length is changed for : #{obj.key}"
    end
    puts "#{obj.key} \t #{obj.content_length}"
  end

  # Move object
  obj = bucket.objects[FileName + ".copy"].move_to(FileName + ".org")
  if !obj.exists?
    raise "File could not Moved Successfully\n"
  end
  puts "\nFile move Successfully\n"

  # List objects in the bucket
  puts "----------List Files---------\n"
  bucket.objects.with_prefix("").each do |obj|
    if !fileObject.size.eql? obj.content_length
      raise " Content length is changed for : #{obj.key}"
    end
    puts "#{obj.key} \t #{obj.content_length}"
  end

  # Rename object
  obj = bucket.objects[FileName + ".org"].rename_to(FileName + ".copy")
  if !obj.exists?
    raise "File could not Rename Successfully\n"
  end
  puts "\nFile rename Successfully\n"

  # List objects in the bucket
  puts "----------List Files---------\n"
  bucket.objects.with_prefix("").each do |obj|
    if !fileObject.size.eql? obj.content_length
      raise " Content length is changed for : #{obj.key}"
    end
    puts "#{obj.key} \t #{obj.content_length}"
  end

  # Download File
  File.open(FileName + ".copy", "w+") do |thisfileObject|
    bucket.objects[FileName].read do |chunk|
      thisfileObject.write(chunk)
    end
    thisfileObject.seek(0)
    thisfileDigest = Digest::MD5.hexdigest(thisfileObject.read)
    if !((thisfileObject.size.eql? metadata.content_length) && (fileDigest.eql? thisfileDigest))
      raise "Downloaded File Metadata could not match"
    else
      puts "\nFile Downloaded Successfully\n"
    end
  end

  # Delete objects one by one and check if exist
  puts "--------------------Delete Files--------------------\n"
  bucket.objects.with_prefix("").each do |obj|
    obj.delete
    puts "#{obj.key} \t File Deleted Successfully..\n"
    if obj.exists?
      raise "Object is not Deleted Successfully\n"
    end
  end

  # Get-Put ACL
  puts "\n#####Default ACL#####"
  puts "Owner ID : #{bucket.acl.owner.id}"
  puts "Owner Display name : #{bucket.acl.owner.display_name}"
  permissions = []
  bucket.acl.grants.each do |grant|
    puts "Bucket ACL is : #{grant.permission.name}"
    puts "Bucket Grantee URI is : #{grant.grantee.uri}"
    permissions << grant.permission.name
  end
  if !(permissions == [:full_control])
    raise "Permission is Not full_control"
  else
    puts "Bucket ACL permission is 'private'\n\n"
  end

  puts "#####:public_read ACL#####"
  bucket.acl = :public_read
  puts "Owner ID : #{bucket.acl.owner.id}"
  puts "Owner Display name : #{bucket.acl.owner.display_name}"
  permissions = []
  bucket.acl.grants.each do |grant|
    puts "Bucket ACL is : #{grant.permission.name}"
    puts "Bucket Grantee URI is : #{grant.grantee.uri}"
    permissions << grant.permission.name
  end
  if !(permissions == [:read, :read_acp] )
    raise "Permission is Not public_read"
  else
    puts "Bucket ACL Successfully changed to 'public-read'\n\n"
  end

  puts "#####:public_read_write ACL#####"
  bucket.acl = :public_read_write
  puts "Owner ID : #{bucket.acl.owner.id} "
  puts "Owner Display name : #{bucket.acl.owner.display_name}"
  permissions = []
  bucket.acl.grants.each do |grant|
    puts "Bucket ACL is : #{grant.permission.name}"
    puts "Bucket Grantee URI is : #{grant.grantee.uri}"
    permissions << grant.permission.name
  end
  if !(permissions == [:read, :read_acp, :write, :write_acp])
    raise "Permission is Not public_read_write"
  else
    puts "Bucket ACL Successfully changed to 'public-read-write'\n\n"
  end

  puts "#####:private ACL#####"
  bucket.acl = :private
  puts "Owner ID : #{bucket.acl.owner.id} "
  puts "Owner Display name : #{bucket.acl.owner.display_name}"
  permissions = []
  bucket.acl.grants.each do |grant|
    puts "Bucket ACL is : #{grant.permission.name}"
    puts "Bucket Grantee URI is : #{grant.grantee.uri}"
    permissions << grant.permission.name
  end
  if !(permissions == [:full_control])
    raise "Permission is Not full_control"
  else
    puts "Bucket ACL Successfully changed to 'private'\n\n"
  end
rescue
  # Unexpected error occurred
  p $!
  exit(-1)
ensure
  # Bucket Delete
  bucket = s3.buckets[Bucket]
  bucket.clear!  #clear the versions only
  bucket.delete
  puts "Bucket deleted Successfully\n"
end
